File: FrontEnd/lib/domain/entities/message.dart
/// @file message.dart
/// @description Entidad de dominio que representa un mensaje de chat.
/// Incluye soporte para actualizaciones optimistas mediante tempId.
/// @module Chat
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// MESSAGE STATUS
// ═══════════════════════════════════════════════════════════════════════════════

/// Estado de entrega de un mensaje a lo largo de su ciclo de vida.
enum MessageStatus {
  /// Enviado localmente, pendiente de confirmación del servidor.
  pending,

  /// Recibido por el servidor (persistido en BD, POST → 201).
  sent,

  /// Entregado al destinatario (recibido vía Socket).
  delivered,

  /// Leído por el destinatario (PUT /chat/:id/read).
  read,
}

// ═══════════════════════════════════════════════════════════════════════════════
// MESSAGE
// ═══════════════════════════════════════════════════════════════════════════════

/// Entidad de dominio de un mensaje de chat.
///
/// El campo [tempId] se asigna en el cliente antes de enviar el mensaje y
/// permite reconciliar actualizaciones optimistas cuando llega la confirmación
/// del servidor (real [id] + estado [delivered]).
class Message {
  /// Identificador único del mensaje (puede ser el tempId mientras es [pending]).
  final String        id;

  /// ID de la conversación a la que pertenece este mensaje.
  final String        conversationId;

  /// ID del usuario que envió el mensaje.
  final String        senderId;

  /// Nombre del usuario que envió el mensaje (enriquecido por el backend).
  final String        senderName;

  /// Contenido textual del mensaje (null en mensajes solo-media).
  final String?       text;

  /// Estado de entrega del mensaje.
  final MessageStatus status;

  /// ID temporal asignado por el cliente para actualizaciones optimistas.
  /// Presente hasta que el servidor confirma el mensaje con su ID real.
  final String?       tempId;

  /// Marca de tiempo de creación del mensaje.
  final DateTime      createdAt;

  const Message({
    required this.id,
    required this.conversationId,
    required this.senderId,
    required this.senderName,
    this.text,
    required this.status,
    this.tempId,
    required this.createdAt,
  });

  // ─── Computed ─────────────────────────────────────────────────────────────

  /// true si el mensaje está pendiente de confirmación del servidor.
  bool get isPending   => status == MessageStatus.pending;

  /// true si el mensaje ha sido recibido por el servidor.
  bool get isSent      => status == MessageStatus.sent;

  /// true si el mensaje ha sido entregado al destinatario.
  bool get isDelivered => status == MessageStatus.delivered;

  /// true si el mensaje ha sido leído por el destinatario.
  bool get isRead      => status == MessageStatus.read;

  /// true si el mensaje tiene contenido de texto.
  bool get hasText     => text != null && text!.isNotEmpty;

  // ─── copyWith ─────────────────────────────────────────────────────────────

  /// Devuelve una copia con los campos indicados modificados.
  ///
  /// Usado principalmente para reconciliar el tempId con el ID real
  /// y actualizar el estado cuando llega el ack del servidor.
  Message copyWith({
    String?        id,
    MessageStatus? status,
    String?        text,
  }) =>
      Message(
        id:             id             ?? this.id,
        conversationId: conversationId,
        senderId:       senderId,
        senderName:     senderName,
        text:           text           ?? this.text,
        status:         status         ?? this.status,
        tempId:         tempId,
        createdAt:      createdAt,
      );
}

File: BackEnd/src/domain/entities/Message.ts
/**
 * @file Message.ts
 * @description Entidad de dominio que representa un mensaje en una conversación de chat.
 * @module Chat
 * @layer Domain
 */

/** Estados posibles de un mensaje en su ciclo de vida de entrega */
export type MessageStatus = 'pending' | 'sent' | 'delivered' | 'read';

/** Metadatos opcionales de contenido multimedia adjunto */
export interface ContentMeta {
  type?: 'image' | 'video' | 'file' | 'audio';
  url?: string;
  size?: number;
  mimeType?: string;
}

/**
 * Entidad de dominio Message.
 *
 * Representa un mensaje enviado por un usuario en una conversación.
 * Soporta texto y contenido multimedia. El tempId permite actualizaciones
 * optimistas en el cliente: se asigna en el lado cliente antes de la confirmación.
 */
export class Message {
  /** Identificador único del mensaje */
  readonly id: string;

  /** ID de la conversación a la que pertenece */
  readonly conversationId: string;

  /** ID del usuario que envió el mensaje */
  readonly senderId: string;

  /** ID del destinatario (para futuras conversaciones de grupo) */
  readonly receiverId?: string;

  /** Texto del mensaje (opcional en mensajes solo-media) */
  readonly text?: string;

  /** Metadatos de contenido multimedia adjunto */
  readonly contentMeta?: ContentMeta;

  /** Estado de entrega del mensaje */
  readonly status: MessageStatus;

  /** ID temporal asignado por el cliente para actualizaciones optimistas */
  readonly tempId?: string;

  /** Fecha de creación del mensaje */
  readonly createdAt: Date;

  /** Fecha de última modificación (actualización de estado) */
  readonly updatedAt: Date;

  constructor(params: {
    id: string;
    conversationId: string;
    senderId: string;
    receiverId?: string;
    text?: string;
    contentMeta?: ContentMeta;
    status: MessageStatus;
    tempId?: string;
    createdAt: Date;
    updatedAt: Date;
  }) {
    this.id             = params.id;
    this.conversationId = params.conversationId;
    this.senderId       = params.senderId;
    this.receiverId     = params.receiverId;
    this.text           = params.text;
    this.contentMeta    = params.contentMeta;
    this.status         = params.status;
    this.tempId         = params.tempId;
    this.createdAt      = params.createdAt;
    this.updatedAt      = params.updatedAt;
  }

  /** true si el mensaje ha sido leído por el destinatario */
  get isRead(): boolean {
    return this.status === 'read';
  }
}
